src/data/markdown/docs/02 javascript api/11 jslib/02 functional.md

```diff
@@ -0,0 +1,38 @@
+---
+title: "functional"
+excerpt: "Functional testing with k6"
+---
+
+The `functional` module is an external JavaScript library that provides APIs to simplify functional testing in k6. 
+This library is inspired by ava, jest and jasmine. 
+
+<Blockquote mod='warning'>
+
+#### This library is rapidly evolving.
+
+This library is stable enough to be useful, but pay attention to the new versions released in jslib.k6.io. 
+
+This documentation is for the last version only. If you discover that some of the code below does not work, it most likely means that you are using an older version.
+
+</Blockquote>
+
+
+See [getting started with functional.js](/javascript-api/jslib/functional/getting-started-with-functional-js) for usage examples.
+
+More advanced examples can be found in the [examples section](/examples/functional-testing)
+
+
+| Function | Description |
+| -------- | ----------- |
+| [test(name, function)](/javascript-api/jslib/functional/test-name-function)  | Entry point for creating tests.  |
+| [expect(value)](/javascript-api/jslib/functional/expect-value)  | expect(value) sets the value to be used in comparison by the next function in the chain |
+| [and(value)](/javascript-api/jslib/functional/and-value)  | and(value) is similar to expect(value), but can be used in a chain. |
+| [as(alias)](/javascript-api/jslib/functional/as-string)  | as(alias) sets a textual representation of the value passed to `expect` or `and`. |
```

Comment:
why is this needed? shouldn't it use the variable name as the textual representation? looks a bit like gold-plating to me.

Reply:
I had the same idea, but sadly in JS you can't use reflection to extract the variable name. I tried to work around it by stringifying the function call and regex out the variable name and that worked :smile:  until it ran through babel and babel minified the code and changed my variable name to `a`.  :sob: 

In other words, it's not possible to extract "response.status" string from code like this `t.expect(response.status).toBe(200)` 

To the solution is to add `.as("response code")` to the chain.

Reply:
Alright. Thanks for the clarification! Babel **should** retain symbol names, however. @MStoykov do you know why this happens? Nothing in https://github.com/loadimpact/k6/blob/a85664b610d11c8b5ab22724bd38cfaa58d84ce3/js/compiler/compiler.go#L96 indicates that it would mangle symbol names, from what I can tell.

Reply:
@simskij I don't think it has anything to do with babel. 
As far as I can understand you both want:
```javascript
var something = 12;
expect(something)
func expect(a) {
  // somehow get that a's name before the call was "something" or "response.status" here
}
```
which seems to *not* be possible in js. 

"The workaround" that is proposed around the internet is to use an object with 1 key that is the name of the variable such as 
```javascript
var something = 1241
function expect(a) {
        console.log(Object.keys(a)[0]);
}
expect({something});
var response = {status:200}

expect({response.status}) // this doesn't work though
```
Unless I am missing something or I misunderstood, this seems like a thing that we won't be able to UX around just with some code and I don't have any better proposals for the API or way to do it as I don't think `as({response_status: response.status})` is at all better so :shrug:

Reply:
Yeah, I might have a look at some point, cause I know other js testing tools do it, but I'm totally fine with how it looks right now. Sorry for opening pandora's box 😂  👍🏼